### ISSUE TITLE
Add weather the instructions came from the system module or the user module while tracing 

### ISSUE THREAD

**Issue by NewAccount777** [mrexodia/dumpulator#66]
![image](https://user-images.githubusercontent.com/125270480/226115852-dec02020-c6e4-49b1-9384-e517832a02d4.png)
Something like that like put the module name on the side. Hope you implement this thanks.

**Comment by mrexodia** [comment]
This is already implemented?

**Comment by NewAccount777** [comment]
> This is already implemented?

Sorry i just saw my bad, 
![image](https://user-images.githubusercontent.com/125270480/226116304-2d52795f-e633-4cdd-abcd-e69f53828030.png)
take a look at this for example, it does not show useful info like where it will return (ret address) like in the system modules or the user modules (user code). if you can implement  what i said above would be nice thanks.

**Comment by mrexodia** [comment]
Yeah, I guess those things make sense to add. So the complete list for this issue would be:

- [ ] Allow only tracing specific regions
- [ ] Print the return address if a `call` instruction is encountered

I only implement features I actually need myself, but you are welcome to give it a shot and make a PR if you choose. The trace function is [`_hook_code`](https://github.com/mrexodia/dumpulator/blob/99fe0bbbdf11f94f2cb946c49084f394135bf463/src/dumpulator/dumpulator.py#L1407).

**Comment by NewAccount777** [comment]
> 

**def _hook_code(uc: Uc, address, size, dp: Dumpulator):
    code = b""
    try:
        code = dp.read(address, min(size, 15))
        instr = next(dp.cs.disasm(code, address, 1))
    except StopIteration:
        instr = None  # Unsupported instruction
    except IndexError:
        instr = None  # Likely invalid memory
    address_name = dp.exports.get(address, "")

    module = ""
    if dp.last_module and address in dp.last_module:
        # same module again
        pass
    else:
        # new module
        dp.last_module = dp.modules.find(address)
        if dp.last_module:
            module = dp.last_module.name

    if address_name:
        address_name = " " + address_name
    elif module:
        address_name = " " + module

    line = f"{hex(address)}{address_name}|"
    if instr is not None:
        line += instr.mnemonic
        if instr.op_str:
            line += " "
            line += instr.op_str
        for reg in _get_regs(instr):
            line += f"|{reg}={hex(dp.regs.__getattr__(reg))}"
        if instr.mnemonic in {"syscall", "sysenter"}:
            line += f"|sequence_id=[{dp.sequence_id}]"
    else:
        line += f"??? (code: {code.hex()}, size: {hex(size)})"

    if address >= <start_address> and address <= <end_address>:
        # write the trace line to a separate file for the desired region
        with open("<region_trace_file>", "a") as region_trace_file:
            region_trace_file.write(line + "\n")

    # write the trace line to the main trace file
    dp.trace.write(line + "\n")**
    
 Would it be something like this? Im sorry,  not totally sure as im limited to my python knowledge, maybe if you see some errors in my code you implement your fixes or just rewrite the entire thing

**Comment by mrexodia** [comment]
The return address if you’re executing a `call` is the next instruction… so cip + insn.size (or whatever capstone uses)

**Comment by regionUser** [comment]
> The return address if you’re executing a `call` is the next instruction… so cip + insn.size (or whatever capstone uses)

Ok problem solved thanks for the help. i should consider learning python more extensively. i also made it more cleaner, so i changed from this:
ret_address_size = 4 //32 bit
ret_address = dp._uc.mem_read(dp.regs.sp, ret_address_size)
to this:
ret_address = uc.reg_read(UC_X86_REG_RIP) + instr.size

def _hook_code(uc: Uc, address, size, dp: Dumpulator):
    code = b""
    try:
        code = dp.read(address, min(size, 15))
        instr = next(dp.cs.disasm(code, address, 1))
    except StopIteration:
        instr = None # Unsupported instruction
    except IndexError:
        instr = None # Likely invalid memory
    address_name = dp.exports.get(address, "")

    module = ""
    if dp.last_module and address in dp.last_module:
        # same module again
        pass
    else:
        # new module
        dp.last_module = dp.modules.find(address)
        if dp.last_module:
            module = dp.last_module.name

    if address_name:
        address_name = " " + address_name
    elif module:
        address_name = " " + module

    line = f"{hex(address)}{address_name}|"
    if instr is not None:
        line += instr.mnemonic
        if instr.op_str:
            line += " "
            line += instr.op_str
        for reg in _get_regs(instr):
            line += f"|{reg}={hex(dp.regs.__getattr__(reg))}"
        if instr.mnemonic == "call":
            # print return address
            ret_address = uc.reg_read(UC_X86_REG_RIP) + instr.size
            line += f"|return_address={hex(ret_address)}"
        if instr.mnemonic in {"syscall", "sysenter"}:
            line += f"|sequence_id=[{dp.sequence_id}]"
    else:
        line += f"??? (code: {code.hex()}, size: {hex(size)})"
    line += "\n"
    dp.trace.write(line)

**Comment by regionUser** [comment]
Man why the format is messed up sorry for that

**Comment by mrexodia** [comment]
You can surround your code with 3x backtick (`)

The RIP is already in the instruction and it’s an argument to the function as well, you don’t have to read it again

**Comment by regionUser** [comment]
Phew done, ok so you want me to create a pr request, or will you just add it from this comment ?

**Comment by regionUser** [comment]
> A pull request will be better. That will run the tests and give a clearer overview of what actually changed.

Is there any time frame for when my pull request will be merged? Thank you. Also, could you please provide me with the location of the tracing function, so that I can try to implement tracing for specific regions? Thank you again.

**Comment by mrexodia** [comment]
No worries, I’ll get to it. The tracing function is the one you already modified…

**Comment by mrexodia** [comment]
I did some initial work on the `flexible-tracing` branch: https://github.com/mrexodia/dumpulator/commit/bbbc3ca7ba6a70c5416e4b63ab769f5dd87c66fe

Currently the idea is that the user derives from the `AbstractTrace` class and can implement their own logic inside of the `step` function. The function `dp.set_tracing()` can be used to enable/disable tracing.

Some thinking is needed about the interface, it would be nice if the user can enable/disable tracing for sequence events (see #27) and also say 'only trace ranges X/Y/Z'.

Unicorn supports tracing specific ranges, but the annoying thing is that the callbacks will be delivered multiple times. This would need some special logic to work as a user might expect.
